### PR TITLE
Feat/dataverse selection filter

### PR DESCRIPTION
**Pull request by 0xLodz**
This PR implements https://github.com/okp4/dataverse-portal/issues/196.

### Implementation details
Before wiring the filters to the ontology catalog, we assume that data passed to `SelectionFilter` components are a list of string. When we get more details about it we would be able to update the component signature accordingly.

#### Autocomplete
image

#### folded filters
image

#### Desktop, light theme
image

#### Checkbox label on 2 lines height w/ ellipsis
image

#### No results fallback
image

### CODE REVIEW

**Review by fredericvilcot — CHANGES_REQUESTED** [review]
Great work, I gave you some comments about the component itself, that maybe could be less coupled to the checkbox mechanism

File: src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+type FilterNameProps = {
+  children: string
+}
+export const FilterName: FC<FilterNameProps> = ({ children }) => (
+  <h3 className="okp4-dataverse-portal-dataverse-filter-name">{children}</h3>
+)
+
+type CheckboxFilterProps = {
+  filterName: string
+  filterValues: string[]
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx:33]: This is related to the `value` attribute of a checkbox element, which for example corresponds to "data analysis", "data management", etc. For these values, `filterName` would be "services categories".

I understand that using "values" is confusing. In order to address this issue, we can probably define a `CheckboxValue` type:

```ts
export type CheckboxValue = string

export type CheckboxOption = {
  checked: boolean
  value: CheckboxValue
  id?: string
  name?: string
  disabled?: boolean
}
```

Then use it for filters:
```tsx
type CheckboxFilterProps = {
  filterName: string
  filterValues: CheckboxValue[]
```

What do you think? Thanks

File: src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  })
+
+  const MAX_SEARCH_RESULTS = 10
+
+  const [searchTerm, setSearchTerm] = useState<string>('')
+
+  const [checkboxOptions, setCheckboxOptions] = useState<CheckboxOption[]>(
+    filterValues
+      .map(value => createDefaultCheckboxOption({ name: filterName, value }))
+      .sort((a, b) => a.value.localeCompare(b.value))
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx:61]: Are you sure that is necessary here? `map` already creates a new array.

File: src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+)
+
+type CheckboxFilterProps = {
+  filterName: string
+  filterValues: string[]
+  searchPlaceholder: string
+}
+
+// eslint-disable-next-line max-lines-per-function
+export const CheckboxFilter: FC<CheckboxFilterProps> = ({
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/checkboxFilter/checkboxFilter.tsx]: Thanks for your feedback. The component has been refactored and renamed with a more generic name: `SelectionFilter`

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    value,
+    selected: false,
+    disabled: false
+  })
+
+  const [searchTerm, setSearchTerm] = useState<string>('')
+
+  const [filterOptions, setFilterOptions] = useState<SelectionOption[]>(
+    filterValues
+      .map(value => createDefaultFilterOption({ name: filterName, value }))
```
Comment (lolottetheclash) [inline src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:67]: Do you really need the `name` in your `filterOption`? as the name is `filterName`, it is available if you need it. Also, I think you don't use your `option` name in the file.

File: src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import type { SelectionOption } from '@/ui/view/dataverse/component/filters/selectionFilter/selectionFilter'
+import './dynamicCheckbox.scss'
+
+type DynamicCheckboxProps = {
+  name: string
+  value: string
+  checked: boolean
+  highlightedTerm: string
+  disabled?: boolean
+  onCheckedChange: (checkboxOption: Pick<SelectionOption, 'selected' | 'value' | 'name'>) => void
```
Comment (lolottetheclash) [inline src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx:15]: I think you can update with an `Omit` ;)

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+type FilterNameProps = {
+  children: string
+}
+export const FilterName: FC<FilterNameProps> = ({ children }) => (
+  <h3 className="okp4-dataverse-portal-dataverse-filter-name">{children}</h3>
+)
+
+type SelectionFilterProps = {
+  filterName: string
+  filterValues: string[]
```
Comment (lolottetheclash) [inline src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:39]: I can see that you sometimes uses `options` to describe `filterValues`, you could have named `filterName`: `label` and `filterValues`: `options`. Not really important, you do as you prefer!

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+  const handleSearch = useCallback(
+    (searchTerm: string) => {
+      setSearchTerm(searchTerm)
+    },
+    [setSearchTerm]
+  )
+
+  const handleSelectedChange = useCallback(
+    (changedOption: Pick<SelectionOption, 'selected' | 'value' | 'name'>) => {
```
Comment (lolottetheclash) [inline src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:79]: same here: I think you can update with an `Omit` ;)

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      setFilterOptions(prevFilterOptions => {
+        const optionIndex = prevFilterOptions.findIndex(
+          option => option.value === changedOption.value
+        )
+        if (optionIndex === -1) {
+          return prevFilterOptions
+        }
+        return [
```
Comment (lolottetheclash) [inline src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:87]: I think this works too:

```
      return prevFilterOptions.map(option =>
          option.value === changedOption.value
            ? { ...option, selected: changedOption.selected }
            : option
        )
```

**Review by lolottetheclash — CHANGES_REQUESTED** [review]
Nice work, I wrote you down some feedbacks 😉

File: src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx (hunk 5 above)
Comment (0xLodz) [reply src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx:15]: The goal of using `Pick` is to explicitly show which props from the checkbox are necessary

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+      <span>{t('noResultsFound')}</span>
+    </div>
+  )
+}
+
+type FilterNameProps = {
+  children: string
+}
+export const FilterName: FC<FilterNameProps> = ({ children }) => (
+  <h3 className="okp4-dataverse-portal-dataverse-filter-name">{children}</h3>
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:34]: I agree, it makes the component cleaner and provide a most appropriate property!

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                          name={filterName}
+                          onCheckedChange={handleSelectedChange}
+                          value={value}
+                        />
+                      )
+                  }
+                })
+              ) : (
+                <div className="okp4-dataverse-portal-selection-filter-no-results-wrapper">
+                  <NoResultsFound />
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:131]: As you rightly pointed out, the `NoResultsFound` component will be reused in the application. Thus we might want to customize it's parent according to the context and probably keep it as is.

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    [filterOptions, searchTerm, maxSearchResults]
+  )
+
+  return (
+    <div className="okp4-dataverse-portal-selection-filter-main">
+      <Collapsible
+        content={
+          <div className="okp4-dataverse-portal-selection-filter">
+            <SearchBar onSearch={handleSearch} placeholder={searchPlaceholder} value={searchTerm} />
+            <div className="okp4-dataverse-portal-selection-filter-list">
```
[the remark this answers is not in the document]
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:111]: Sounds a more appropriate name!

File: src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx (hunk 5 above)
Comment (lolottetheclash) [reply src/ui/view/dataverse/component/dynamicCheckbox/dynamicCheckbox.tsx:15]: Works for me !

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx (hunk 4 above)
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:67]: You're right, we don't need it to filter the options. The name props will be necessary when the filter is linked to the ontology.

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx (hunk 6 above)
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:39]: Yes `label` might be an appropriate naming too. The`filterName` naming was chosen to match the `name` props of the `SelectionOption`.
The Filter options is more like a `SelectionOption`:
```tsx
type SelectionOption = {
    selected: boolean;
    value: string;
    name: string;
    disabled: boolean;
}
```

File: src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx (hunk 8 above)
Comment (0xLodz) [reply src/ui/view/dataverse/component/filters/selectionFilter/selectionFilter.tsx:87]: Yes it works too and it's less verbose. We can update it if you consider it easier to read.

**Review by fredericvilcot — APPROVED** [review]

**Review by lolottetheclash — APPROVED** [review]
Nice work !

File: src/ui/page/dataverse/dataverse.tsx [hunk 12]
```diff
@@ -433,6 +438,36 @@ const dataverseItems: DataverseItemDetails[] = [
   }
 ]
 
+const selectionFilters = {
```
Comment (ccamel) [inline src/ui/page/dataverse/dataverse.tsx:441]: These data should be fetched from the ontology, as it contains the controlled vocabularies and is therefore the unique source of truth. Will they be taken into account soon?

Comment (0xLodz) [reply src/ui/page/dataverse/dataverse.tsx:441]: Yes, the wiring to the ontology is carried out in the next iteration, in a dedicated PR.

**Pull request merged**